File: src/main/java/pl/yalgrin/playnite/simplesync/config/JsonMapperProvider.java
package pl.yalgrin.playnite.simplesync.config;

import com.fasterxml.jackson.annotation.JsonInclude;
import lombok.AccessLevel;
import lombok.NoArgsConstructor;
import org.springframework.boot.jackson.autoconfigure.JsonMapperBuilderCustomizer;
import tools.jackson.databind.json.JsonMapper;

@NoArgsConstructor(access = AccessLevel.PRIVATE)
public class JsonMapperProvider {
    public static JsonMapper build() {
        JsonMapper.Builder builder = JsonMapper.builder();
        customizer().customize(builder);
        return builder.build();
    }

    public static JsonMapperBuilderCustomizer customizer() {
        return builder -> {
            builder.disable(tools.jackson.databind.DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES);
            builder.changeDefaultPropertyInclusion(value -> value.withValueInclusion(JsonInclude.Include.NON_NULL));
            builder.changeDefaultPropertyInclusion(
                    value -> value.withContentInclusion(JsonInclude.Include.NON_DEFAULT));
        };
    }
}

File: src/main/kotlin/pl/yalgrin/playnite/simplesync/common/config/JsonMapperProvider.kt
package pl.yalgrin.playnite.simplesync.common.config

import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonSetter
import com.fasterxml.jackson.annotation.Nulls
import org.springframework.boot.jackson.autoconfigure.JsonMapperBuilderCustomizer
import tools.jackson.databind.DeserializationFeature
import tools.jackson.databind.json.JsonMapper

fun buildJsonMapper(): JsonMapper {
    return JsonMapper.builder().customize().build()
}

fun JsonMapper.Builder.customize(): JsonMapper.Builder {
    JsonMapperBuilderCustomizer { builder ->
        builder.disable(
            DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES,
            DeserializationFeature.FAIL_ON_NULL_FOR_PRIMITIVES
        )
        builder.changeDefaultPropertyInclusion { value ->
            value.withValueInclusion(
                JsonInclude.Include.NON_NULL
            )
        }
        builder.changeDefaultPropertyInclusion { value -> value.withContentInclusion(JsonInclude.Include.NON_DEFAULT) }
        builder.withConfigOverride(List::class.java) { o -> o.setNullHandling(JsonSetter.Value.forValueNulls(Nulls.AS_EMPTY)) }
    }.customize(this)

    return this
}